180-adhoc-activities

Buckwich opened the pull request:
Closes #180

### Proposed Changes

<!--

Add relevant context (issue fixed or related to), 
a capture of the UI changes (if any) as well as 
steps to try out your changes.

--> 

### Checklist

To ensure you provided everything we need to look at your PR:

* [ ] **Brief textual description** of the changes present
* [ ] **Visual demo** attached
* [ ] **Steps to try out** present, i.e. [using the `@bpmn-io/sr` tool](https://github.com/bpmn-io/sr)
* [ ] Related issue linked via `Closes {LINK_TO_ISSUE}` or `Related to {LINK_TO_ISSUE}`

<!--

Thanks for creating this pull request! ❤️

-->

nikku reviewed (commented):
Thanks for this PR, and the comprehensive visual description of what it does.

One thing that both of you discussed is whether it should only "disable for activities" or any flow-nodes, which is this case you attached, right?

![image](https://github.com/user-attachments/assets/d2bd898d-c6bc-4c90-b665-3208f384d674)

Ad-hoc sub-processes are special, because start events and disconnected gateways are forbidden anyway, and elements are fine ("meant to be") disconnected there. Using [bpmn-js-bpmnlint](https://github.com/bpmn-io/bpmn-js-bpmnlint) I validated this new rule against the `recommended` bpmnlint settings, some feedback:

To my knowledge, implicit start does not exist in the context of an ad-hoc sub-process. Elements must always be explicity activated.

![image](https://github.com/user-attachments/assets/37687bb1-a1b8-473d-9bed-c8eaa7af19df)

Implicit ends do not exist inside an ad-hoc sub-process. It is completed by the user once "all work is done", whatever that means.

![image](https://github.com/user-attachments/assets/aeb85783-834f-4145-b75c-1e995e7ae7a0)

`no-disconnected` also generally does not make sense, within that element. We do have `ad-hoc-sub-process` as a rule that validates strictly connection semantics within an ad-hoc sub-process.

In this light, I propose to simplify your solution:

* The `no-disconnected`, `no-implicit-end` and `no-implicit-start` should be disabled for all flow nodes within an ad-hoc sub process

nikku commented:
We also have to be careful to return `false` from a rule evaluation, because [it will prevent entering the sub-tree to check nested nodes](https://github.com/bpmn-io/bpmnlint/blob/main/lib/traverse.js#L25).

nikku commented on rules/no-implicit-end.js:
```diff
@@ -63,6 +63,10 @@ module.exports = function() {
     }
 
+    if (is(node, 'bpmn:Activity') && is(node.$parent, 'bpmn:AdHocSubProcess')) {
+      return false;
```
Must not be false, cf. https://github.com/bpmn-io/bpmnlint/pull/181#issuecomment-2739396363

barmac replied:
Let's not confuse this helper function `isImplicitEnd` with the `check` function that linting rules provide. This helper is totally fine to return `false` value as it is used only to decide on `if`: https://github.com/bpmn-io/bpmnlint/pull/181/files#diff-f6e66cae9b06c698bf93b455d0b8d8993a5dff0a8d14672389a6dc2a565f72caR91-R92

The "return to stop traversing" applies only in the `check` method, so that's why we return undefined in https://github.com/bpmn-io/bpmnlint/pull/181/files#diff-f6e66cae9b06c698bf93b455d0b8d8993a5dff0a8d14672389a6dc2a565f72caR88

tldr: I think Nico's suggestions regarding "must not be false" don't apply in this PR.

To be absolutely sure, @Buckwich you could a test case with nested subprocess:

![image](https://github.com/user-attachments/assets/851f368f-5696-494d-b211-545b17fcc078)

nikku replied:
> tldr: I think Nico's suggestions regarding "must not be false" don't apply in this PR.

My appologies, if I misread part of the source code. Great you double-checked @barmac :star_struck:

nikku commented on rules/no-implicit-start.js:
```diff
@@ -29,6 +29,10 @@ module.exports = function() {
     }
 
+    if (is(node, 'bpmn:Activity') && is(node.$parent, 'bpmn:AdHocSubProcess')) {
+      return false;
```
Must not be `false`, cf. https://github.com/bpmn-io/bpmnlint/pull/181#issuecomment-2739396363.

Buckwich replied:
not a direct return of `check`. cf https://github.com/bpmn-io/bpmnlint/pull/181#discussion_r2005537986

nikku commented on rules/no-implicit-start.js:
```diff
@@ -29,6 +29,10 @@ module.exports = function() {
+    }
+
     if (is(node, 'bpmn:SubProcess') && node.triggeredByEvent) {
       return false;
```
Must not be false, I think, cf. https://github.com/bpmn-io/bpmnlint/pull/181#issuecomment-2739396363

nikku commented on rules/no-implicit-end.js:
```diff
@@ -63,6 +63,10 @@ module.exports = function() {
+    }
+
     if (is(node, 'bpmn:EndEvent')) {
       return false;
```
It is misleading that this is false, cf. https://github.com/bpmn-io/bpmnlint/pull/181#issuecomment-2739396363.

barmac commented:
> implicit start does not exist in the context of an ad-hoc sub-process 
> Implicit ends do not exist inside an ad-hoc sub-process

I agree with these statements.

> `no-disconnected` also generally does not make sense, within that element. We do have `ad-hoc-sub-process` as a rule that validates strictly connection semantics within an ad-hoc sub-process.

Yes, but [the rule](https://github.com/bpmn-io/bpmnlint/blob/main/docs/rules/ad-hoc-sub-process.md) does not apply to the gateways. Perhaps most suitable solution would be to disable `no-disconnected` for ad-hoc subprocess, and implement gateway validation in `ad-hoc-sub-process` (both incoming and outgoing required).

nikku commented:
> Perhaps most suitable solution would be to disable `no-disconnected` for ad-hoc subprocess, and implement gateway validation in ad-hoc-sub-process (both incoming and outgoing required).

Yes, I'd disable that rule. No, I'd not add dedicated validation for gateways, unless a user reports that as an issue. In other words, the following is an absolutely fine use of an ad-hoc sub-process to me. You can decide whether you want that "complex decision" to be taken, or any other decision, at any point in time:

![image](https://github.com/user-attachments/assets/27cd2e19-e59c-47ef-ab16-7f77149505fa)

barmac commented:
Sure we can always wait for more feedback. Still, the main problem to solve is that activities are marked as disconnected/implicit-start/implicit-end where they should not be marked (false positive).

Buckwich commented:
I've adjusted the rules to always skip if in an adhoc-subprocess. I've kept returning false. And removed the invalid test case as they are now valid. I'll probably also squash all commits into one as it is always the same line of codes

nikku reviewed (approved):
Simple enough, thanks for following up! :star: 

Let's squash and merge things.
